### PR TITLE
Cosmology clone

### PR DESCRIPTION
**Pull request by rainwoodman**

### CODE REVIEW

**Comment by rainwoodman** [comment]
This changes some of the API -- as cosmology instance is made immutable (since usually makes the logic of serialization etc simpler)

I find it a bit strange LinearPower doesn't `match` sigma8 of the cosmology, but HaloFit does. I am not sure if this is intended.

**Comment by rainwoodman** [comment]
This will close #382

**Comment by rainwoodman** [comment]
The only down side is Cosmology(**dict(c)) won't work, but it shall be Cosmology.from_dict(dict(c)) -- given that `__init__` doesn't strongly suggest a dict as input it is probably not too bad of a compromise for avoiding keeping both args and pars.

File: nbodykit/cosmology/cosmology.py [hunk 1]
```diff
@@ -35,8 +37,10 @@ class Cosmology(object):
       inferred by the curvature condition.
     * Non-cosmological constant dark energy can be used by specifying the
       ``w0_fld``, ``wa_fld``, and/or ``Omega_fld`` values.
-    * The ``sigma8`` attribute can be set to the desired value, and the internal
-      value of ``A_s`` will be automatically adjusted.
+    * To pass in CLASS parameters that are not valid Python argument names, use
+      the dictionary /keyward arguments trick, e.g.
+      `Cosmology(..., **{'temperature contributions': 'y'})`
+    * ``Cosmology(**dict(c)) is not supposed to work; use ``Cosmology.from_dict(dict(c))``.
```
Comment (nickhand) [inline nbodykit/cosmology/cosmology.py]: Can you use double ticks for the Cosmology(..., **{'temperature contributions': 'y'}) line? And close the ticks on Cosmology(**dict(c)) so the docs look okay?

File: nbodykit/cosmology/cosmology.py [hunk 2]
```diff
@@ -10,6 +10,8 @@ class Cosmology(object):
     A cosmology calculator based on the CLASS binding in :mod:`classylss`.
 
     It is a collection of all method provided by the CLASS interfaces.
+    The object is immutable. To obtain an instance with a new set of parameters
+    use :method:`clone` or :method:`match`.
```
Comment (nickhand) [inline nbodykit/cosmology/cosmology.py]: Should we add an explicit __setattr__ that raises an Exception? I think you can currently set attributes, although it doesn't do anything.

**Comment by nickhand** [comment]
For LinearPower, sigma8 just re-scales the amplitude, so we can don't need to keep re-computing the ClassEngine when sigma8 changes. And we also want the sigma8 attribute to agree with the sigma_r(8) function at z=0.

**Comment by nickhand** [comment]
We should also fix k=0 in linear power so it doesn't raise a true_divide warning

**Comment by nickhand** [comment]
merging this into a bccp:cosmology-clone branch

**Comment by nickhand** [comment]
see #386

**Pull request merged**